### PR TITLE
Fix compiler errors on Windows

### PR DESCRIPTION
**Pull request by yocabon**
## What does this PR do?
Fixes # (issue).

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/xformers/blob/master/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - https://github.com/facebookresearch/xformers/issues/437
- [x] Did you make sure to update the docs?
  - [x] N/A
- [x] Did you write any new necessary tests?
  - [x] N/A
- [x] Did you update the [changelog](https://github.com/facebookresearch/xformers/blob/master/CHANGELOG.md)? (if needed)
  - [x] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -113,6 +113,7 @@ def get_flash_attention_extensions(cuda_version: int, extra_compile_args):
                 "nvcc": extra_compile_args.get("nvcc", [])
                 + [
                     "-O3",
+                    "-std=c++17",
```
Comment (fmassa) [inline setup.py:116]: Can this be conditionned on Windows builds only?

File: xformers/components/attention/csrc/attention.cpp [hunk 2]
```diff
@@ -1,5 +1,18 @@
...
+// For PyMODINIT_FUNC to work, we need to include Python.h
+// https://github.com/pytorch/vision/blob/main/torchvision/csrc/vision.cpp#L17
+// Fixes error LNK2001: unresolved external symbol PyInit__C
+#if defined(_WIN32)
+#include <Python.h>
+PyMODINIT_FUNC PyInit__C(void) {
+  // No need to do anything.
+  return NULL;
+}
+#endif // defined(_WIN32)
```
Comment (fmassa) [inline xformers/components/attention/csrc/attention.cpp:14]: Thanks for this! This is a longstanding issue with torchvision as well which we haven't found a better solution so far.

File: xformers/components/attention/csrc/cuda/mem_eff_attention/kernel_backward.h [hunk 3]
```diff
@@ -1348,12 +1348,14 @@ struct AttentionBackwardKernel {
...
+    // on windows, previous syntax __restrict__ AccessType*
+    // resulted in error: "restrict" is not allowed 
+    const AccessType* __restrict__ grad_output_ptr =
+        reinterpret_cast<const AccessType* __restrict__>(
             p.grad_output_ptr + (query_start + laneRow) * p.head_dim_value +
             laneFirstCol);
-    const __restrict__ AccessType* output_ptr =
-        reinterpret_cast<const __restrict__ AccessType*>(
+    const AccessType* __restrict__ output_ptr =
+        reinterpret_cast<const AccessType* __restrict__>(
```
Comment (fmassa) [inline xformers/components/attention/csrc/cuda/mem_eff_attention/kernel_backward.h:1358]: @danthe3rd looks like we might have switched the order of the `restrict` keywords here. Do you think the `__restrict__` keyword was being picked by the compiler before in here?

**Review by fmassa — APPROVED** [review]
This all LGTM, thanks a ton @yocabon !

I only have really one request -- could you try making C++17 be a dependency only on Windows for flashattention?

File: setup.py (hunk 1 above)
Comment (yocabon) [reply setup.py]: https://github.com/facebookresearch/xformers/pull/460/commits/bf5b1fd922bec9a7723723e1d014e7d09a60fcea should do it

File: xformers/components/attention/csrc/cuda/mem_eff_attention/attention_backward_generic.cu [hunk 4]
```diff
@@ -139,13 +139,23 @@ mem_efficient_attention_backward_cutlass(
...
+    // };
+    // TORCH_INTERNAL_ASSERT(
+    //     checkBinaryArchMatches(),
+    // "Something went wrong in the build process");
+    cudaFuncAttributes attr;
+    AT_CUDA_CHECK(cudaFuncGetAttributes(&attr, kernel_fn));
     TORCH_INTERNAL_ASSERT(
-        checkBinaryArchMatches(), "Something went wrong in the build process");
+        attr.binaryVersion >= Kernel::ArchTag::kMinComputeCapability,
+        "Something went wrong in the build process");
```
Comment (danthe3rd) [inline xformers/components/attention/csrc/cuda/mem_eff_attention/attention_backward_generic.cu:158]: as this is just a sanity check - can we keep it as it was but under an `#ifdef` guard for windows?

Comment (yocabon) [reply xformers/components/attention/csrc/cuda/mem_eff_attention/attention_backward_generic.cu]: https://github.com/facebookresearch/xformers/pull/460/commits/8509baaef125121c4cf7f5476eb4c26e1d0769c6 should do it

File: xformers/components/attention/csrc/cuda/mem_eff_attention/kernel_backward.h (hunk 3 above)
Comment (danthe3rd) [reply xformers/components/attention/csrc/cuda/mem_eff_attention/kernel_backward.h]: that's a good point. I'll run some benchmarks after that

**Review by danthe3rd — APPROVED** [review]
Thanks a lot! That's amazing!
Let's wait for the CI before merging. I believe there might be a merge conflict also ...

**Pull request merged**